test/metabase/lib/datascript_test.cljc

```diff
@@ -0,0 +1,1100 @@
+                (ld/visible-column-groups query -1))))
+
+      (testing "pieces in latter stage"
+        (let [[created-at
+               count-col] (->> query :mbql.query/stage0 :mbql.stage/returned (sort-by :mbql/series))
+              filtered    (ld/filter query -1 (ld/expr :> count-col 100))]
+          (is (=? {:mbql.query/stage
+                   {:mbql.stage/filter
+                    [{:mbql.clause/operator :>
+                      :mbql.clause/argument [{:mbql/series 0
+                                              :mbql/ref    (->> filtered
+                                                                :mbql.query/stage0
+                                                                :mbql.stage/returned
+                                                                (sort-by :mbql/series)
+                                                                last)}
+                                             {:mbql/series 1, :mbql/lit 100}]
+                      :mbql/series          0}]}}
+                  filtered)))))))
+
+(defn- transitive-closure
+  [edges seen x]
+  (let [reachable (for [edge edges
+                        :let [targets (when-let [t (get x edge)]
+                                        (if (set? t) t #{t}))]
+                        :when targets
+                        target targets
+                        :when (not (seen target))]
+                    target)]
+    (reduce (partial transitive-closure edges) (apply conj seen reachable) reachable)))
+
+(defn- no-dangling-refs
+  "Searches for any ref attributes pointing at an entity that no longer exists."
+  [query]
+  (let [db         (d/entity-db query)
+        referenced (d/q '[:find [?ref ...]
+                          :in $ [?ref-attr ...] :where
+                          [_ ?ref-attr ?ref]]
+                        db ld/ref-attrs)]
+    (is (empty? (filter #(nil? (d/datoms db :eavt %)) referenced)))))
+
+(defn- no-orphaned-entities
+  "Searches for entities which still exist in the DB but are not either reachable from the query, or exist independently
+  like the metadata."
+  [query]
+  (let [;; Metadata - databases, tables, fields, cards - exist independently.
+        bedrock-attrs [:metadata.database/id :metadata.table/id :metadata.field/id :metadata.card/id]
+        db            (d/entity-db query)
+        ;; `bedrock` is the set of independently real things - metadata, etc.
+        bedrock       (set (d/q '[:find [?bedrock ...]
+                                  :in $ [?bedrock-attr ...] :where
+                                  [?bedrock ?bedrock-attr]]
+                                db bedrock-attrs))
+        entities      (into #{} (map :e) (d/datoms db :eavt))
+        reachable     (into #{} (map :db/id) (transitive-closure ld/ref-attrs #{query} query))]
+    (is (empty? (remove (set/union bedrock reachable) entities)))))
+
+(deftest ^:parallel remove-clause-test-basic-soundness
```

Comment:
And now a bunch of permutations of `remove-clause`, with nontrivial cascades.

I need to write loads more of these for confidence, but the foundation is sound.